#138 5 Tips to Write Better Conditionals in JavaScript (#139)
jobbole/translation-project: translation/#138-5-Tips-to-Write-Better-Conditionals-in-JavaScript.md

> ## 5 Tips to Write Better Conditionals in JavaScript（用 JavaScript 写出更好条件语句的 5 个技巧）

Reviewer: 题目我觉得可以改改，优化javascript条件语句的5个技巧

> ## 5 Tips to Write Better Conditionals in JavaScript（用 JavaScript 写出更好条件语句的 5 个技巧）

Author: ```suggestion
## 5 Tips to Write Better Conditionals in JavaScript（优化 JavaScript 条件语句的5个技巧）
```

> 转译自：https://scotch.io/tutorials/5-tips-to-write-better-conditionals-in-javascript

When working with JavaScript, we deal a lot with conditionals, here are the 5 tips for you to write better / cleaner conditionals.

在使用 JavaScript 时，我们会处理很多条件语句，这里有 5 个技巧可以帮助您编写更好、更简洁的条件语句。

### 1、Use Array.includes for Multiple Criteria（对多个条件使用 Array.includes）

Let's take a look at the example below:

让我们看看下面的例子：

```
// condition
function test(fruit) {
  if (fruit == 'apple' || fruit == 'strawberry') {
    console.log('red');
  }
}
```

At first glance, the above example looks good. However, what if we get more red fruits, say cherry and cranberries? Are we going to extend the statement with more || ?

> 乍一看，上面的例子看起来不错。然而，如果我们得到更多的红色水果，比如樱桃和小红莓呢？我们要用更多的 || 来扩展这个表述吗？

Reviewer: 如果我们得到更多水果，这里不看原文就有点突兀，因为get翻译为得到不妥

如果还有更多类型的水果需要判断呢

> 乍一看，上面的例子看起来不错。然而，如果我们得到更多的红色水果，比如樱桃和小红莓呢？我们要用更多的 || 来扩展这个表述吗？

Author: ```suggestion
乍一看，上面的例子看起来不错。然而，如果还有更多红颜色的水果需要判断呢，比如樱桃和小红莓，我们要用更多的 || 来扩展这个表述吗？
```

  const redFruits = ['apple', 'strawberry', 'cherry', 'cranberries'];

  if (!fruit) throw new Error('No fruit!'); // condition 1: throw error early
  if (!redFruits.includes(fruit)) return; // condition 2: stop when fruit is not red

  console.log('red');

  // condition 3: must be big quantity
  if (quantity > 10) {
    console.log('big quantity');
  }
}
```

By inverting the conditions of condition 2, our code is now free of a nested statement. This technique is useful when we have long logic to go and we want to stop further process when a condition is not fulfilled.

通过反转条件 2 的条件，我们的代码现在没有嵌套语句。当我们有很长的逻辑要处理时，这种技术是有用的，当一个条件没有满足时，我们想要停止进一步的处理。

However, that's no **hard rule** for doing this. Ask yourself, is this version (without nesting) better / more readable than the previous one (condition 2 with nested)?

然而，这并不是**严格的规则**。问问自己，这个版本（没有嵌套）是否比前一个版本（嵌套的条件 2）更好、更易读？

For me, I would just leave it as the previous version (condition 2 with nested). It is because:

对于我来说，我将把它保留为以前的版本（条件 2 和嵌套）。这是因为：

- the code is short and straight forward, it is clearer with nested if

代码简短而直接，如果嵌套，代码就更清晰了

- inverting condition may incur more thinking process (increase cognitive load)

> 反转条件可能会导致更多的思考过程（增加认知负荷）

Author: ```suggestion
反转条件可能会导致更多的思考过程（增加认知负担）
```

  if (!fruit) return;
  console.log(`We have ${quantity} ${fruit}!`);
}

//test results
test('banana'); // We have 1 banana!
test('apple', 2); // We have 2 apple!
```

Much easier & intuitive isn't it? Please note that each parameter can has it own [default function parameter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters). For example, we can assign default value to `fruit` too: `function test(fruit = 'unknown', quantity = 1)`.

更简单和直观，不是吗？请注意，每个参数都可以有自己的默认函数参数。例如，我们也可以为 `fruit` 赋值：`function test(fruit = 'unknown'， quantity = 1)`。

What if our `fruit` is an object? Can we assign default parameter?

如果我们的 `fruit` 是一个对象呢？我们可以指定默认参数吗？

```
function test(fruit) {
  // printing fruit name if value provided
  if (fruit && fruit.name)  {
    console.log (fruit.name);
  } else {
    console.log('unknown');
  }
}

//test results
test(undefined); // unknown
test({ }); // unknown
test({ name: 'apple', color: 'red' }); // apple
```

Look at the example above, we want to print the fruit name if it's available or we will print unknown. We can avoid the conditional fruit && fruit.name checking with default function parameter & destructing.

> 看看上面的例子，我们想打印水果名，如果它是可用的，或者我们将打印 unknown。我们可以避免使用与默认函数参数和解构对条件 fruit && fruit.name 进行检查。

Reviewer: 如果它是可用的，这里需要调整语序

> 看看上面的例子，我们想打印水果名，如果它是可用的，或者我们将打印 unknown。我们可以避免使用与默认函数参数和解构对条件 fruit && fruit.name 进行检查。

Author: ```suggestion
请看上面的示例，如果 `fruit.name` 是可用的，我们将打印该水果名称，否则我们将打印 `unknown`。我们可以避免使用与默认函数参数和解构对条件 `fruit && fruit.name` 进行检查。
```

This last tip is more about utilizing new (but not so new) Javascript Array function to reduce the lines of code. Look at the code below, we want to check if all fruits are in red color:

最后一个技巧是关于使用新的（但不是很新）Javascript 数组函数来减少代码行。看看下面的代码，我们想检查所有的水果是否都是红色的:

```
const fruits = [
    { name: 'apple', color: 'red' },
    { name: 'banana', color: 'yellow' },
    { name: 'grape', color: 'purple' }
  ];

function test() {
  let isAllRed = true;

  // condition: all fruits must be red
  for (let f of fruits) {
    if (!isAllRed) break;
    isAllRed = (f.color == 'red');
  }

  console.log(isAllRed); // false
}
```

The code is so long! We can reduce the number of lines with `Array.every`:

代码太长了！我们可以用 `Array.every` 来减少行数：

```
const fruits = [
    { name: 'apple', color: 'red' },
    { name: 'banana', color: 'yellow' },
    { name: 'grape', color: 'purple' }
  ];

function test() {
  // condition: short way, all fruits must be red
  const isAllRed = fruits.every(f => f.color == 'red');

  console.log(isAllRed); // false
}
```

Much cleaner now right? In a similar way, if we want to test if any of the fruit is red, we can use `Array.some` to achieve it in one line.

> 现在干净多了，对吧？类似地，如果我们想测试任何一个水果是红色的，我们可以使用 `Array.some` 来实现它在一行。

Reviewer: 用一行语句实现它

> 现在干净多了，对吧？类似地，如果我们想测试任何一个水果是红色的，我们可以使用 `Array.some` 来实现它在一行。

Author: ```suggestion
现在干净多了，对吧？类似地，如果我们想用一行代码来判断任何一个水果是否为红色，我们可以使用 `Array.some`。
```
